track_analyzing/track_matcher.cpp

```diff
@@ -0,0 +1,254 @@
+uint64_t constexpr kShortTrackDuration = 60;
+
+double DistanceToSegment(m2::PointD const & segmentBegin, m2::PointD const & segmentEnd,
+                         m2::PointD const & point)
+{
+  m2::ProjectionToSection<m2::PointD> projection;
+  projection.SetBounds(segmentBegin, segmentEnd);
+  m2::PointD const projectionPoint = projection(point);
+  return MercatorBounds::DistanceOnEarth(point, projectionPoint);
+}
+
+double DistanceToSegment(Segment const & segment, m2::PointD const & point, IndexGraph & indexGraph)
+{
+  return DistanceToSegment(indexGraph.GetGeometry().GetPoint(segment.GetRoadPoint(false)),
+                           indexGraph.GetGeometry().GetPoint(segment.GetRoadPoint(true)), point);
+}
+
+bool EdgesContain(vector<SegmentEdge> const & edges, Segment const & segment)
+{
+  for (auto const & edge : edges)
+  {
+    if (edge.GetTarget() == segment)
+      return true;
+  }
+
+  return false;
+}
+}  // namespace
+
+namespace track_analyzing
+{
+// TrackMatcher ------------------------------------------------------------------------------------
+TrackMatcher::TrackMatcher(storage::Storage const & storage, NumMwmId mwmId,
+                           platform::CountryFile const & countryFile)
+  : m_mwmId(mwmId)
+  , m_vehicleModel(CarModelFactory().GetVehicleModelForCountry(countryFile.GetName()))
+{
+  auto localCountryFile = storage.GetLatestLocalFile(countryFile);
+  CHECK(localCountryFile, ("Can't find latest country file for", countryFile.GetName()));
+  auto registerResult = m_index.Register(*localCountryFile);
+  CHECK_EQUAL(registerResult.second, MwmSet::RegResult::Success,
+              ("Can't register mwm", countryFile.GetName()));
+
+  m_graph = make_unique<IndexGraph>(
+      GeometryLoader::Create(m_index, registerResult.first, m_vehicleModel),
+      EdgeEstimator::CreateForCar(nullptr /* trafficStash */, m_vehicleModel->GetMaxSpeed()));
+
+  MwmSet::MwmHandle const handle = m_index.GetMwmHandleByCountryFile(countryFile);
+  DeserializeIndexGraph(*handle.GetValue<MwmValue>(), *m_graph);
+}
+
+void TrackMatcher::MatchTrack(vector<DataPoint> const & track, vector<MatchedTrack> & matchedTracks)
+{
+  m_pointsCount += track.size();
+
+  vector<Step> steps;
+  steps.reserve(track.size());
+  for (auto const & routePoint : track)
+    steps.emplace_back(routePoint);
+
+  for (size_t trackBegin = 0; trackBegin < steps.size();)
+  {
+    for (; trackBegin < steps.size(); ++trackBegin)
+    {
+      steps[trackBegin].FillCandidatesWithNearbySegments(m_index, *m_graph, *m_vehicleModel,
+                                                         m_mwmId);
+      if (steps[trackBegin].HasCandidates())
+        break;
+
+      ++m_nonMatchedPointsCount;
+    }
+
+    if (trackBegin >= steps.size())
+      break;
+
+    size_t trackEnd = trackBegin;
+    for (; trackEnd < steps.size() - 1; ++trackEnd)
+    {
+      Step & nextStep = steps[trackEnd + 1];
+      nextStep.FillCandidates(steps[trackEnd], *m_graph);
```

Comment:
Please, add comment to steps[trackEnd]. It's quite misleading here, at least add comment that it's a previous step, please.